pennylane/capture/capture_transforms.py

```diff
@@ -0,0 +1,318 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Trace implementations for program capture"""
+from typing import Any, Optional
+
+import jax
+import jax.numpy as jnp
+from jax.core import MainTrace, Primitive, ShapedArray, Trace, Tracer
+
+import pennylane as qml
+
+from .base_interpreter import PlxprInterpreter
+from .primitives import adjoint_transform_prim, ctrl_transform_prim
+
+_mp_return_types = tuple(repr(value) for value in qml.measurements.ObservableReturnTypes)
+
+
+class TransformTrace(Trace):
+    """Trace for processing primitives for PennyLane transforms.
+
+    Args:
+        main (jax.core.MainTrace): Active interpreter to which this ``Trace`` belongs.
+        sublevel (int): Sublevel for this ``Trace``.
+        transform_program (pennylane.transforms.core.TransformProgram): ``TransformProgram``
+            containing the transforms to apply to the primitives being interpreted.
+        state (Optional[dict]): Dictionary containing environment information for transforms
+            to use.
+    """
+
+    def __init__(
+        self,
+        main: MainTrace,
+        sublevel: int,
+        transform_program: "qml.transforms.core.TransformProgram",
+        state: Optional[dict] = None,
+    ):
+        super().__init__(main, sublevel)
+        self._transform_program = transform_program
+        self._state = state or {}
+
+    def pure(self, val: Any):
+        """Create a new ``TransformTracer``."""
+        return TransformTracer(self, val, 0)
+
+    lift = sublift = pure
+
+    @property
+    def state(self):
+        """Dictionary containing environment information for transforms to use."""
+        return self._state
+
+    def process_primitive(self, primitive: Primitive, tracers: tuple[Tracer], params: dict):
+        """Interpret a given primitive.
+
+        Args:
+            primitive (jax.core.Primitive): Primitive to interpret.
+            tracers (Sequence[jax.core.Tracer]): Input tracers to the primitive.
+            params (dict): Keyword arguments/metadata for the primitive.
+
+        Returns:
+            Any: The result of the interpretation.
+        """
+        idx = max(t.idx for t in tracers if isinstance(t, TransformTracer))
+        is_qml_primitive = (
+            primitive.__class__.__module__.split(".")[0] == "pennylane"
+            or primitive.name.split("_")[0] in _mp_return_types
```

Comment:
So measurements don't have the module name of "pennylane"?

Reply:
Also, what happens if someone is using a custom operator? Should we have a backlog task to figure that out?

Reply:
> So measurements don't have the module name of "pennylane"?

No, because measurement primitives are `jax.core.Primitive`s so their namespace is `jax`, but the rest of the primitives in pennylane are `pennylane.capture.NonInterpPrimitive`s. We could try to make the measurement primitives `NonInterpPrimitive`s as well, but I feel like that is not soemthing that should be done in this PR.

Reply:
> Also, what happens if someone is using a custom operator? Should we have a backlog task to figure that out?

Good point, it will not work out of the box. I considered standardizing the primitive names by adding `qml.` to the front of the primitive names, but eventually reverted that change, because that also seems like it should be done in a different PR.

Reply:
cc @dwierichs in hindsight, why aren't measurement primitives not `NonInterpPrimitive`s?

Reply:
@dwierichs For context, the `NonInterpPrimitive` is the new name for `NonJVPPrimitive` (or a similar name) you created some time ago. I re-named it after intercepting `BatchTracers` along with `JVPTracers` as well

Reply:
Thanks for the context, @PietropaoloFrisoni 

@mudit2812 I just added it where it was needed from the standpoint of the epic requirements, basically.
I suppose if we want to differentiate through measurements, that actually will currently break due to them not being `NonInterpPrimitive`s (?)
There was no inherent reason to _not_ make them a custom Primitive.